Add laws.json

jcushman opened the pull request:
Here's a first pass at adding `laws.json` based on the state-by-state rules from the [Indigo Book](https://law.resource.org/pub/us/code/blue/IndigoBook.html#T3). Offering for discussion!

See below for the full file, but this initial proposal is basically a few hundred entries that look like this:

```
    {
        "cite_type": "leg_statute",
        "end": null,
        "examples": [
            "Ala. Code § 92.979 (1999)"
        ],
        "jurisdiction": "Alabama",
        "name": "Code of Alabama, 1975 (West)",
        "notes": "Automatically generated on 2021-04-30. Remove this message if replacing with real examples.",
        "regexes": [
            "Ala\\. Code § $statute_section \\($statute_year\\)"
        ],
        "start": null
    }
```

Open questions:

* Do we like putting all the non-court legislative and executive stuff in "laws.json"?
* Top level structure: this is a list of dicts, with no top-level keys. It could have a structure like `jurisdiction` or `jurisdiction -> cite_type` or something else? Is there anything analogous to the reporters.json structure that's useful here?
* JSON keys:
  * `cite_type`: currently has this totally arbitrary mapping from the Indigo Book categories:
    ```
    "Statutory compilation": "leg_statute",
    "Session laws": "leg_session",
    "Uncompiled law": "leg_act",
    "Administrative compilation": "admin_compilation",
    "Administrative register": "admin_register",
    "Administrative and executive register": "admin_register",
    "Municipal regulations": "municipal",
    ```
  * `start` and `end`: currently `null`. I'm not interpreting the occasional date ranges from the Indigo Book as being restrictive, since those might just be the preferred citation date range, so only adding them to the notes field. Probably valid date ranges are still useful though?
  * `jurisdiction`: currently the full state name, rather than e.g. an `mlz_jurisdiction` value.
  * `regexes` and `examples`: should work the same way as reporters.json, with placeholders from regexes.json. There's nothing analogous to `$edition` substitution, so each entry results in one regex. Examples are automatically generated for now.
  * Missing fields?
* Choices in converting citation forms to regular expressions:
  * Here's the substitutions I made to generate the regular expressions, followed by some thoughts:
  ```
    # agency/subject/department-name
    "<agency abbreviation>": r"(?P<agency>\w+)",  # wisconsin agency
    "<subject>": "$statute_subject",
    "<Subject>": "$statute_subject",
    "<law>": "$statute_subject",
    "<dep't name as abbreviated in Rules>": "$statute_subject",

    # title/article/chapter/number/page/volume/...
    "tit. x": r"tit\. (?P<title>\d+)",
    "art. x": r"art (?P<article>\d+)",
    "ch. x,": r"ch\. (?P<chapter>\d+),",
    "x-x-x Me.": r"(?P<chapter>\d+-\d+-\d+) Me\.",
    "x-x-x Del.": r"(?P<chapter>\d+-\d+-\d+) Del\.",
    "x-x-x Ark.": r"(?P<chapter>\d+-\d+-\d+) Ark\.",
    "<ch. x § x>": r"ch\\. (?P<chapter>\d+) § $statute_section",
    "<ch. no.>": r"(?P<chapter>\d+)",
    "<page no.>": "$page_with_commas",
    "<vol. no.>": "$volume",
    "<pamph. no.>": r"(?P<pamphlet>\d+)",
    "<tit. no.>": r"(?P<title>\d+)",
    "<act no.>": r"(?P<act>\d+)",
    "<iss. no.>": r"(?P<issue>\d+)",
    "<law no.>": r"(?P<law>\d+)",
    "<rule no.>": r"(?P<rule>\d+)",
    "<reg. no.>": r"(?P<reg>\d+)",
    "<reg no.>": r"(?P<reg>\d+)",
    "No. x": r"No\. (?P<number>\d+)",

    # section
    "§ x:x-x.x": "§ $statute_section",
    "§ x:x-x-x": "§ $statute_section",
    "§ x.x.x.x": "§ $statute_section",
    "§ x-x-x-x": "§ $statute_section",
    "§ x.x.x": "§ $statute_section",
    "§ x-x-x": "§ $statute_section",
    "§ x-x.x": "§ $statute_section",
    "§ x.x": "§ $statute_section",
    "§ x-x": "§ $statute_section",
    "§ x:x": "§ $statute_section",
    "§ x": "§ $statute_section",
    "Regs. x-x-x.x": r"Regs\. $statute_section",
    "r. x-x-x.x": r"r\. $statute_section",
    "r. x.x.x.x": r"r\. $statute_section",
    "r. x-x.x": r"r\. $statute_section",
    "r. x-x-x": r"r\. $statute_section",
    "r. x.x": "r. $statute_section",
    "<sec. no.>": "$statute_section",

    # dates
    "<year>": "$statute_year",
    "<month year>": "$statute_month $statute_year",
    "<month day, year>": "$statute_month $statute_day $statute_year",
    ```
  * Thoughts:
    * There isn't consistency in capture group naming here like "volume", "reporter", "year" that we have in reporters.json. I'm not sure how these want to be used downstream, but maybe there could be more consolidated naming for groups like number, reg, rule, law, issue, act, title, pamphlet.
    * I didn't match all the `section` groups with their own pattern, but used one pattern with up to four digit groups separated by one of `:-.`.
    * I didn't look into whether all of these that say `\d` are actually `\d`. Acts could be roman numerals, for example; others could have commas or other formatting. I'm thinking when we get this running we could then do a report for regexes that have the fewest matches and see what's broken.

Known incomplete in this demo, but could be fixed before pulling or done later:

* No imports or helpers yet, just the raw json.
* No tests yet (e.g. to ensure regexes match).
* No federal law yet. This is just the state-level data from section T3.

mlissner commented:
Interesting stuff. My general inclination is to be super liberal in what we do here, and just see what works. The general outline you have above seems about right to me. I do feel like it has perhaps more inconsistency with reports.json than might be ideal, but:

1. `mlz_jurisdictions` isn't really used much that I know of, so it's probably fine to take a different approach for it. In any case, statutes are less granular than courts, so it seems fine to simplify. Perhaps @fbennett would have thoughts about how important that is here though. (Actually, Frank, if you have thoughts about this in general, I'm sure we'd greatly benefit.)

2. I miss the `variations` dicts, but they're not really needed either if we have `examples` and `regexes`. 

3. I too would sort of like to have it be a dictionary, just for consistency sake, but I agree there's no obvious key for that.

So yeah. Maybe it's fine?

For the big question you've got here, "How do we split up our regexes?" I can think of two use cases:

1. People often ask something like, "Can you give me all RECAP cases that involve patent XYZ or statute ABC." It'd be nice to handle at least the statute part of that. I guess patents can come ~someday.

2. At least a simple goal of this would be to make useful links, but I have no idea the right granularity to do that. Maybe Sara Frug from LII would have thoughts? I'd be happy to chat with her. I'm overdue. 

Whatever the answers to the above, I feel fine merging this once tests pass, though those three todos you listed would be nice. 

Finally: Holy cow, man!

jcushman commented:
> I miss the variations dicts

I wasn't sure if variations served a purpose when there isn't a consistent concept of "editions" here, but one thing we could do eventually is make it more generic, so "variations" provides arbitrary placeholders for the regexes. Like given this example:

```
        "regexes": [
            "Ala\\. Code § $statute_section \\($statute_year\\)"
        ],
```

if we later realized there were a bunch of variations for "Ala. Code", we could change it to:

```
        "regexes": [
            "$code § $statute_section \\($statute_year\\)"
        ],
        "variations": {
            "code": ["Ala\\. Code", "Ala\\.", "Ala\\. Foo"]
        }
```

And a consumer like eyecite can generate all the regex variants from that. On the other hand that could just be `(?:Ala\\. Code|Ala\\.|Ala\\. Foo)` in the original regex, so maybe it's not needed.

mlissner commented:
That's an interesting approach. I guess it has nice locality improvements over our current approach. I'm not sure it's really worth it though to have two places to look for regexes, or worth it to have a different approach with one json file vs another. I'd say keep it simple for now, myself, but not a strong feeling though.

jcushman commented:
I think this might be ready to go, for a first pass. I added `from reporters_db import LAWS` and refactored the tests to factor out some helpers that can be applied to laws.json as well as reporters.json: check_whitespace, check_ascii, check_dates, check_json_format, check_regexes. Like you said, let's stay loose and expect that the format of this file might want to change.

jcushman commented:
Actually hold on, I think I might be able to extract useful keys from all of these, like

```
Virgin Islands Code Annotated Advance 'V.I. Code Ann. Adv.'
Legislative Service (LexisNexis) 'Legis. Serv.'
Code of U.S. Virgin Islands Rules (LexisNexis) 'V.I. Code R.'
Virgin Islands Government Register (LexisNexis) "V.I. Gov't Reg."
```

mlissner commented:
My review is complete. I'm ready to merge once you give the thumbs up on any data issues.

jcushman commented:
OK, it turns out I can get decent short keys for these by stripping out capture groups from the regular expressions and keeping what's left (plus stripping section indicators like `tit\.|§|r\.|art\.?|,|No\.|ch\.|/`). What's left looks like this, a little closer to reporters.json:

```
    "Ala. Admin. Code": [
        {
            "cite_type": "admin_compilation",
            "end": null,
            "examples": [
                "Ala. Admin. Code r. 218"
            ],
            "jurisdiction": "Alabama",
            "name": "Alabama Administrative Code",
            "notes": "Automatically generated. Remove this message if replacing with real examples.",
            "regexes": [
                "$edition r\\. $law_section"
            ],
            "start": null,
            "variations": []
        }
    ],
```

So users like eyecite can substitute `(?<reporter>Ala. Admin. Code|variations...)` in for `$edition`, just like we do with reporters.json. The terms "edition" and "reporter" aren't quite right but I kept them for compatibility since nothing else is that great either.

This gets a little funky for like three reporters where the subject area of the law is embedded inside the reporter string, like `"Tex. Code Ann."` has the regex `"(?P<reporter>Tex\\. $law_subject Code Ann\\.) § $law_section"`. Those three won't be able to use variations; oh well.

I also stripped out publisher and date suffixes in this update -- these ones:

```
    " \\(Consol\\. $law_year\\)",
    " \\(McKinney $law_year\\)",
    " \\(Deering $law_year\\)",
    " \\($law_month $law_year\\)",
    " \\(LexisNexis $law_month $law_year\\)",
    " \\(LexisNexis $law_year\\)",
    " \\(LexisNexis\\)",
    " \\($law_month $law_day $law_year\\)",
    " \\(West $law_year\\)",
    " \\(West\\)",
    " \\($law_year\\)",
```

So we no longer have two entries for things like "1999 Colo. Sess. Laws 1" and "1999 Colo. Sess. Laws 1 (West)", but just combine them into one without the suffix. It makes more sense to add the parenthetical as cite metadata in eyecite rather than making it a mandatory part of the pattern here.

mlissner commented:
Not gonna lie, the familiarity of that json file feels good. Were there any instances where mapping the abbreviation to a list of dicts was necessary? Would mapping the key straight to a dict be better/simpler?

jcushman commented:
For the two that are currently lists, they could probably be handled with a broader regex or with two regexes, rather than with two entries -- it's just an optional pamphlet number:

* R.I. Adv. Legis. Serv.:
  * Rhode Island Advance Legislative Service (LexisNexis): <year>-<pamph. no.> R.I. Adv. Legis. Serv. <page no.>
  * West’s Rhode Island Advance Legislative Service: <year> R.I. Adv. Legis. Serv. <page no.>
* Ga. Code Ann. Adv. Legis. Serv.:
  * Georgia Advance Legislative Service (LexisNexis): <year>-<pamph. no.> Ga. Code Ann. Adv. Legis. Serv. <page no.>
  * West’s Georgia Legislative Service: <year> Ga. Code Ann. Adv. Legis. Serv. <page no.>

I'm imagining we might dig up more troublesome ones as we go along, though? Like this is just Indigo Book stuff where everything has been carefully sorted out and has a state included, but I know I'll be adding a "G.L. c." variation for internal Massachusetts statutory cites at some point, and I could see internal citation stuff like that start to collide more, especially ones (unlike Mass statutes) that don't have an official version. 🤷‍♂️
